### PR TITLE
Utilize arithmetic operators to deal with Permissions/Intents

### PR DESCRIPTION
**Pull request by HopeBaron**
This PR aims to:
* Replace `PermissionsBuilder` and `IntentsBuilder` with simple arithmetic operators
* Introduce a vararg invoke function to feed in multiple `Permission`s and/or `Intents`

### CODE REVIEW

File: common/src/main/kotlin/com/gitlab/kordlib/common/entity/Permission.kt [hunk 1]
```diff
@@ -2,51 +2,99 @@ package com.gitlab.kordlib.common.entity
...
     }
 
-    @OptIn(ExperimentalContracts::class)
-    inline fun copy(block: PermissionsBuilder.() -> Unit): Permissions {
-        contract {
-            callsInPlace(block, InvocationKind.EXACTLY_ONCE)
-        }
-        val builder = PermissionsBuilder(code)
-        builder.apply(block)
-        return builder.permissions()
```
Comment (BartArys) [inline common/src/main/kotlin/com/gitlab/kordlib/common/entity/Permission.kt:37]: Why did we end up removing this? I thought it was quite useful for scoping more complex operations.

File: common/src/main/kotlin/com/gitlab/kordlib/common/entity/Permission.kt [hunk 2]
```diff
@@ -2,51 +2,99 @@ package com.gitlab.kordlib.common.entity
...
     override fun toString(): String {
         return "Permissions(values=${values.joinToString(", ")})"
     }
 
     companion object : KSerializer<Permissions> {
 
-        inline operator fun invoke(block: PermissionsBuilder.() -> Unit = {}): Permissions {
-            val builder = PermissionsBuilder()
-            builder.apply(block)
-            return builder.permissions()
```
Comment (BartArys) [inline common/src/main/kotlin/com/gitlab/kordlib/common/entity/Permission.kt:49]: Same here, why did we end up removing this?

File: core/src/main/kotlin/com/gitlab/kordlib/core/builder/kord/KordBuilder.kt [hunk 3]
```diff
@@ -128,28 +125,6 @@ class KordBuilder(val token: String) {
...
-     * }
-     * ```
-     */
-    @OptIn(ExperimentalContracts::class)
-    inline fun intents(builder: Intents.IntentsBuilder.() -> Unit) {
-        contract {
-            callsInPlace(builder, InvocationKind.EXACTLY_ONCE)
-        }
-        intents = Intents { builder() }
-    }
```
Comment (BartArys) [inline core/src/main/kotlin/com/gitlab/kordlib/core/builder/kord/KordBuilder.kt:152]: Same question here, why did we end up removing this?

File: core/src/main/kotlin/com/gitlab/kordlib/core/entity/Member.kt [hunk 4]
```diff
@@ -90,17 +90,13 @@ class Member(
...
+        if (owner) return Permissions(Permission.All)
 
         val everyone = guild.getEveryoneRole().permissions
         val roles = roles.map { it.permissions }.toList()
 
-        return Permissions {
-            +everyone
-            roles.forEach { +it }
-        }
+        return roles.fold(everyone) { acc, permissions -> acc + permissions }
```
Comment (BartArys) [inline core/src/main/kotlin/com/gitlab/kordlib/core/entity/Member.kt:98]: This is quite a bit less efficient than the previous version. Obviously this won't lead to any noticeable differences but I think it's a good argument in favour of not removing the builder DSL.

File: core/src/main/kotlin/com/gitlab/kordlib/core/entity/channel/GuildChannel.kt [hunk 5]
```diff
@@ -52,28 +52,25 @@ interface GuildChannel : Channel, GuildChannelBehavior {
...
-
+        roleOverwrites.forEach {
+            result += it.allowed
+            result -= it.denied
+        }
+        if (memberOverwrite != null) {
+            result += memberOverwrite.allowed
+            result -= memberOverwrite.denied
+        }
+        return result
```
Comment (BartArys) [inline core/src/main/kotlin/com/gitlab/kordlib/core/entity/channel/GuildChannel.kt:73]: This also seems like a good argument for not deleting the DSL, I find this less readable and less performant.

File: gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt [hunk 6]
```diff
@@ -149,83 +152,97 @@ enum class Intent(val code: Int) {
...
+    operator fun minus(intent: Intent): Intents = Intents(code xor intent.code)
 
     /**
      * Returns an [Intents] that removed the [intent] from this [code].
      */
-    operator fun minus(intent: Intent): Intents = when {
-        code and intent.code == intent.code -> Intents(code xor intent.code)
-        else -> this
+
+    operator fun minus(intent: Intents): Intents = Intents(code xor intent.code)
```
Comment (BartArys) [inline gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt:182]: This is incorrect; see above.

File: common/src/main/kotlin/com/gitlab/kordlib/common/entity/Permission.kt [hunk 7]
```diff
@@ -2,51 +2,99 @@ package com.gitlab.kordlib.common.entity
...
+     * Returns a new [Permissions] that removes [permission] from its [code]
+     */
 
     operator fun minus(permission: Permission): Permissions = Permissions(code xor (code and permission.code))
 
+    /**
+     * Returns a new [Permissions] that removes [permission] from its [code]
+     */
+
+    operator fun minus(permission: Permissions): Permissions = Permissions(code xor (code and permission.code))
```
Comment (BartArys) [inline common/src/main/kotlin/com/gitlab/kordlib/common/entity/Permission.kt:38]: extra parentheses are unnecessary if you switch the two sides of either operation

File: common/src/main/kotlin/com/gitlab/kordlib/common/entity/Permission.kt [hunk 8]
```diff
@@ -2,51 +2,99 @@ package com.gitlab.kordlib.common.entity
...
+     * Returns a new [Permissions] that adds [permission] to its [code]
+     */
+
+    operator fun plus(permission: Permissions): Permissions = Permissions(this.code or permission.code)
+
+    /**
+     * Returns a new [Permissions] that removes [permission] from its [code]
+     */
 
     operator fun minus(permission: Permission): Permissions = Permissions(code xor (code and permission.code))
```
Comment (BartArys) [inline common/src/main/kotlin/com/gitlab/kordlib/common/entity/Permission.kt:32]: extra parentheses are unnecessary if  you switch the two sides of either operation

File: gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt [hunk 9]
```diff
@@ -149,83 +152,97 @@ enum class Intent(val code: Int) {
...
+     * removes this intent from the [code] of an [Intents]
      */
-    @OptIn(ExperimentalContracts::class)
-    inline fun copy(block: IntentsBuilder.() -> Unit): Intents {
-        contract {
-            callsInPlace(block, InvocationKind.EXACTLY_ONCE)
-        }
-        val builder = IntentsBuilder(code)
-        builder.apply(block)
-        return builder.flags()
```
Comment (BartArys) [inline gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt:184]: Same here, why did we end up removing this?

File: gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt [hunk 10]
```diff
@@ -149,83 +152,97 @@ enum class Intent(val code: Int) {
...
+
+    /**
+     * Returns an [Intents] that added [code]s  in [Intents] to this [code].
+     */
+    operator fun plus(intent: Intents): Intents = Intents(this.code or intent.code)
+
+    /**
+     * Returns an [Intents] that removed the [intent] from this [code].
+     */
+    operator fun minus(intent: Intent): Intents = Intents(code xor intent.code)
```
Comment (BartArys) [inline gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt:176]: This is incorrect:

Intents(0B1100) - Intent(0B0010):
```
1100
0010
------- XOR
1110 <--expected--> 1100
```
If you want to inline this function into a single expression you should use `Intents XOR Intent AND Intents` instead:
```
1100
0010
------- XOR
1110 
1100 (Intents)
-------- AND
1100
```
use case where intent is in intents:
```
1100
0100
------- XOR
1000 
1100 (Intents)
-------- AND
1000
```

File: gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt [hunk 11]
```diff
@@ -149,83 +152,97 @@ enum class Intent(val code: Int) {
...
+
+    operator fun minus(intent: Intents): Intents = Intents(code xor intent.code)
+
+
+
+
+    /**
+     * adds this intent to the [code] of an [Intents]
+     */
+
```
Comment (BartArys) [inline gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt:190]: newline

File: gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt [hunk 12]
```diff
@@ -149,83 +152,97 @@ enum class Intent(val code: Int) {
...
+
     companion object {
 
         @PrivilegedIntent
         val all: Intents
-            get() = invoke {
-                Intent.values().forEach { +it }
-            }
+            get() = Intents(*Intent.values())
```
Comment (BartArys) [inline gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt:226]: The spread operator is all kinds horrible in terms of performance: https://sites.google.com/a/athaydes.com/renato-athaydes/posts/kotlinshiddencosts-benchmarks

consider using the integer constructor instead here, or adding a Iterable constructor.

File: rest/src/main/kotlin/com/gitlab/kordlib/rest/builder/channel/ChannelPermissionModifyBuilder.kt [hunk 13]
```diff
@@ -13,7 +13,7 @@ class ChannelPermissionModifyBuilder(private var type: String) : AuditRequestBui
     /**
      * The permissions that are explicitly allowed for this channel.
      */
-    var allowed: Permissions = Permissions()
+    var allowed: Permissions =   Permissions()
```
Comment (BartArys) [inline rest/src/main/kotlin/com/gitlab/kordlib/rest/builder/channel/ChannelPermissionModifyBuilder.kt:16]: unnecessary space

File: gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt [hunk 14]
```diff
@@ -149,83 +152,97 @@ enum class Intent(val code: Int) {
...
 
-                -Intent.GuildPresences
-                -Intent.GuildMembers
-            }
+        val none: Intents = Intents()
 
-        val none: Intents = invoke()
-
-        inline operator fun invoke(builder: IntentsBuilder.() -> Unit = {}): Intents {
-            return IntentsBuilder().apply(builder).flags()
```
Comment (BartArys) [inline gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt:207]: Why did we end up removing this?

File: gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt [hunk 15]
```diff
@@ -149,83 +152,97 @@ enum class Intent(val code: Int) {
...
+        operator fun invoke(vararg intents: Intent): Intents {
```
Comment (BartArys) [inline gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt:234]: Consider adding an Iterable variant as well.

File: common/src/main/kotlin/com/gitlab/kordlib/common/entity/Permission.kt [hunk 16]
```diff
@@ -2,51 +2,99 @@ package com.gitlab.kordlib.common.entity
...
         return "Permissions(values=${values.joinToString(", ")})"
     }
 
     companion object : KSerializer<Permissions> {
 
-        inline operator fun invoke(block: PermissionsBuilder.() -> Unit = {}): Permissions {
-            val builder = PermissionsBuilder()
-            builder.apply(block)
-            return builder.permissions()
+        operator fun invoke(vararg permissions: Permission): Permissions {
```
Comment (BartArys) [inline common/src/main/kotlin/com/gitlab/kordlib/common/entity/Permission.kt:89]: Consider adding an Iterable variant as well.

File: common/src/main/kotlin/com/gitlab/kordlib/common/entity/Permission.kt [hunk 17]
```diff
@@ -2,51 +2,99 @@ package com.gitlab.kordlib.common.entity
...
+            val result = permissions.fold(0) { acc, permission -> acc or permission.code }
+            return Permissions(result)
+        }
+
+
+        operator fun invoke(vararg permissions: Permissions): Permissions {
```
Comment (BartArys) [inline common/src/main/kotlin/com/gitlab/kordlib/common/entity/Permission.kt:95]: Consider adding an Iterable variant as well.

File: gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt [hunk 18]
```diff
@@ -149,83 +152,97 @@ enum class Intent(val code: Int) {
...
-
-        operator fun Intent.unaryPlus() {
-            this@IntentsBuilder.code = this@IntentsBuilder.code or code
-        }
 
-        operator fun Intent.unaryMinus() {
-            if (this@IntentsBuilder.code and code == code) {
-                this@IntentsBuilder.code = this@IntentsBuilder.code xor code
-            }
+        operator fun invoke(vararg intents: Intents): Intents {
```
Comment (BartArys) [inline gateway/src/main/kotlin/com/gitlab/kordlib/gateway/Intent.kt:240]: Consider adding an Iterable variant as well.

**Review by BartArys — CHANGES_REQUESTED** [review]
Lot's of small things related to markup, try to stay consistent with capitalizing the start of the sentence and adding punctuation add the end. 

There's also the removal of the builders which I don't agree with. There's some examples in the new code that I've commented on that should make a case for keeping them.

File: core/src/main/kotlin/com/gitlab/kordlib/core/entity/channel/GuildChannel.kt (hunk 5 above)
Comment (HopeBaron) [reply core/src/main/kotlin/com/gitlab/kordlib/core/entity/channel/GuildChannel.kt:73]: This is was the idea behind making `code` property mutable. I also considered making it internal as well. I guess I still need to work around the bikeshidding habit of mine.

**Review by BartArys — APPROVED** [review]

**Review by BartArys — APPROVED** [review]
I've got major issues with this code and I can't check if it actually works since CI is failing, but I've been told it works.

**Pull request merged**